Implements rank system for Torch (code help wanted)

Kasuobes opened the pull request:
See commit comment. Will require assistance to remove the disgusting amount of copypasta. This is what it looks like right now (using SGR and RD only as examples, jobs not fully configured):

![](http://i.imgur.com/4YkPbCL.png)

Kasuobes committed:
Implements rank system for Torch

Service branch and rank is now selectable in character setup under
occupation. Allowed ranks can be configured for individual jobs. If the
selected rank isn't allowed for a specific job the job will become
unavailable for selection.

Raptor1628 commented:
Shouldn't civilian/none allow research jobs? In any case, can I get a rundown on the features? It locks out jobs, yes, but does it also lock out alt-titles? Or Custom Loadout stuff? Does it spawn specific ranks?

Kasuobes commented:
- Research jobs are not available in the screenshot because I'm only using SGR and RD for testing. 
- Alt titles are tied to the jobs, if the job isn't available the alt title shouldn't pop up.
- Insignia and uniforms can be configured independently of the job outfits. I suppose it's possible to utilize pref.char_branch and pref.char_rank to selectively blacklist/whitelist items in custom loadout.

In the simplest sense, characters can now have branch and rank selected. Pretty much anything you can think of that should be using the branch and rank system can be done by expanding the code in the future, such as giving characters with specific ranks and/or in specific branches their own access, independent of their jobs.

This scope of this PR is just to implement the branch/rank system that allow restrictions to jobs though.

Raptor1628 commented:
Okay, so if you were to say select "Fleet" "Corporal," and join as an approved job, it would automatically override the job's loadout to add the uniform and appropriate rank pin?

Kasuobes commented:
The fleet branch wouldn't have corporal within the list of available ranks. When outfits are configured for rank/branch they should be able to override the outfits for jobs.

Raptor1628 commented:
Okay, so it does also lock out ranks based on branch? This all looks very well done.

Kasuobes commented:
Rank gets locked out based on branch, yes.

Kasuobes commented:
Added example configs to torch_jobs.dm.

SubatomicTeddybear commented:
Sounds/looks nifty, :+1: 

Also looks like it works, then again, I'm no coder.

Kasuobes commented:
It works, but the code is disgusting.

SubatomicTeddybear commented:
@Kasuobes All code is disgusting, but some is more disgusting than others. Your code is nice,
_"<NobleCaos> Don't fix Noble code, Noble code isn't good for you mental or physical health"_

NobleCaos commented on maps/torch/torch_occupation.dm:
```diff
@@ -0,0 +1,184 @@
+		if(RETURN_TO_LOBBY)
+			. += "<u><a href='?src=\ref[src];job_alternative=1'>Return to lobby if preference unavailable</a></u>"
+
+	. += "<a href='?src=\ref[src];reset_jobs=1'>\[Reset\]</a></center>"
```
ONly thing that bugs me is that little red thing around the "/] ", but I assume that's just git being pissy.

SubatomicTeddybear commented:
`+  . += "<a href='?src=\ref[src];reset_jobs=1'>\[Reset\]</a></center>"`

There's a red thing around the `\]` but I assume that's just github being github.

Raptor1628 commented:
Forgot to ask, an rank/branch allow/deny certain custom loadout options?

Kasuobes commented:
The items might have to be categorized for whitelist/blacklist, but it should be fairly straightforward.

BlueNexus commented:
Could someone expand this later on to make this system automatically assign uniforms and rank accessories? I know the system is capable, this is a request for someone to do so.

Raptor1628 commented:
@BlueNexus it seems like that's how it works already.

Kasuobes commented:
Will need help to make the code look less ugly, I'm hitting the limits of my coding knowledge, can't figure out how to make datums work.

SubatomicTeddybear commented:
@Kasuobes Need help with the map datum?

Kasuobes commented:
Trying to consolidate the massive #define list into datums, failing miserably.

Also still have no idea how to reduce copy/paste code in torch_occupation.dm.

PsiOmegaDelta commented on maps/torch/torch_occupation.dm:
```diff
@@ -0,0 +1,184 @@
+			pref.char_rank = choice
+			return TOPIC_REFRESH
+
+	return ..()
```
As maps can be rotated, you cannot have unique loading/saving per map, or data will be lost on map change.

comma replied:
Not sure about that, since savefiles are not recreated, maybe 'unknown' fields will be left untouched. Better test that by running save proc WITH ranks, save WITHOUT ranks and load WITH ranks and see if it fucked up.

Kasuobes replied:
Resolved by refactoring rank/branch so that the preferences are global, but the option to change them can be togglable per map.

Kasuobes committed:
Refactors branch/rank system for map dependency

Branch/rank can now be enabled/disabled in map settings. Also solves the
problem of having potential unknown fields in save files.

Kasuobes commented:
Branch/rank can now be enabled/disabled in map settings. Also solves the problem of having potential unknown fields in save files.

Also included better checks for job.allowed_ranks and job.allowed_branches, if a job don't have the lists populated it is assumed there's no rank/branch restriction.

Kasuobes committed:
Rewrote list of ranks

Ranks and branches are now vars in their own datums instead of tons of #DEFINE

mustafakalash commented on maps/torch/torch_jobs.dm:
```diff
@@ -75,6 +87,7 @@
+	allowed_ranks = list(CIV_CIVILIAN)
 
 /datum/job/contractor
 	title = "Security Contractor"
```
Name was changed to Security Guard

Raptor1628 replied:
Just leave job crap for now, it's my next major project.

mustafakalash commented:
Besides my comment, is this finished?

Kasuobes commented:
It's ugly, but it works.

GinjaNinja32 commented on maps/torch/torch_occupation.dm:
```diff
@@ -0,0 +1,46 @@
+		else if(pref.char_branch == SCG_MARINE)
+			choice = input(user, "Choose your rank.", "Character Preference", pref.char_rank) as null|anything in SCG_MARINE
+		else if(pref.char_branch == SCG_CIVILIAN)
+			choice = input(user, "Choose your rank.", "Character Preference", pref.char_rank) as null|anything in SCG_CIVILIAN
```
is `char_branch` here intended to be being compared to things that are apparently lists?

if this is right, perhaps:

```
var/options = pref.char_branch
if(options == SCG_EXPCORP)
    options = SCG_FLEET

var/choice = input(user, "Choose your rank.", "Character Preference", pref.char_rank) as null|anything in options
```

GinjaNinja32 replied:
Reading further, they're strings. This won't work.

GinjaNinja32 commented on maps/torch/torch_ranks.dm:
```diff
@@ -0,0 +1,76 @@
+
+//some ranks not meant for public usage, can be VV'd for adminbus
+var/datum/ranks
+	var/const/CIV_CIVILIAN = "Civilian"
```
What's with the weird indentation here?

Kasuobes committed:
Fixes rank selection, removed pointless datums

GinjaNinja32 commented on maps/~mapsystem/maps.dm:
```diff
@@ -59,6 +58,9 @@ var/list/all_maps = list()
 
+	var/HAS_BRANCH	//Branch system for occupations, togglable
+	var/HAS_RANK	//Rank system, also togglable
+
```
``` dm
var/const/MAP_HAS_BRANCH = 1
var/const/MAP_HAS_RANK = 2

/datum/map
    var/flags = 0
```

``` dm
/datum/map/torch
    flags = MAP_HAS_BRANCH | MAP_HAS_RANK
```

GinjaNinja32 commented on maps/torch/torch_ranks.dm:
```diff
@@ -0,0 +1,74 @@
+var/const/FLEET_O8 = "Rear Admiral"
+var/const/FLEET_O9 = "Vice Admiral"
+var/const/FLEET_O10 = "Admiral"
+var/const/FLEET_O10_ALT = "Fleet Admiral"
```
These could almost certainly do with being lists rather than a lot of consts, especially with the `MARINE_*` following the same structure.

Kasuobes committed:
Resolve conflict with torch_define.dm, add flags

Still WIP

Kasuobes committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into torch

Kasuobes commented:
Now that my finals are done, I can finally sit down and work on this properly. Lemme see if I can do this...

Kasuobes committed:
Merge branch 'dev' into torch

Kasuobes commented:
Still working.

Kasuobes commented:
Because of how my lack of understanding in lists is holding back evertything, it might be best to merge this in its current usable form and clean it up later. I'm really not satisfied though.

@comma >.>

Finalsong commented:
@Kasuobes your PR is beautiful, do not feel down

SubatomicTeddybear commented:
https://media.giphy.com/media/MpHGtb8s1gwow/giphy.gif

Looks nice to me, @Kasuobes

PsiOmegaDelta commented on maps/torch/torch_occupation.dm:
```diff
@@ -0,0 +1,46 @@
+//Shamelessly copied over form occupation.dm, needs cleaning up
+/datum/category_item/player_setup_item/occupation/OnTopic(href, href_list, user)
```
Coming across no real issues except this monster of a copy-paste..
This definitely needs to be generalized in such a way that it can be contained in at the original location and re-used by any other maps that might want to use the rank system.

Kasuobes committed:
Further reduced copy/paste in torch_occupations.dm

It works, surprisingly.

Kasuobes commented:
Addressed monster copy-paste. Somehow it now works even though it didn't before for me.

/shrug.

Finalsong commented:
yay!

BlueNexus commented:
All we need now is an automatic uniform assignment system.

Kasuobes commented:
Outfit datums, set rank outfits to override job outfits, or have one load after the other and don't override if a gear slot is already assigned.

BlueNexus commented:
Needs to assign the full outfit though, including utes, service and dress uniform, all the hats they would need, rank stuff, etc. They can then put what they don't need in a locker.
